### PR TITLE
br: add new design for batch create table

### LINKED ISSUE #27036: Speed up the DDL execution speed of TiDB backup data restore
## Feature Request

**Is your feature request related to a problem? Please describe:**

**Describe the feature you'd like:**

The cluster has 6 TiB data, 30k tables, and 11 TiKVs. When I use BR to backup and restore the cluster, I find that the speed is particularly slow. After investigation, BR can only create 2 tables per second, the entire store speed takes nearly 4h, and the execution time spent creating tables is close to 4h. It can be seen that the execution speed of ddl is the bottleneck in the this scenario.

**Describe alternatives you've considered:**

In order to speed up the restore speed, I hope TiDB to speed up the execution of DDL. In this scenario, BR can create tables at 200 tables/s
 In terms of compatibility, Binlog/CDC is required to be able to replicate the table schema of these tables, and the restored data does not need to be replicated temporarily

- [x] #30272
- [x] #30284
- [x] #30516
- [ ] CDC can sync data between upstream (TiDB) to downstream

### PR DESCRIPTION
**Pull request by fengou1**
### What problem does this PR solve?
add new design for batch create table

Issue Number: close #27036

Problem Summary:

### Check List

Tests 

NA

Side effects

- [ ] Performance regression: Consumes more CPU
- [ ] Performance regression: Consumes more Memory
- [ ] Breaking backward compatibility

Documentation

- [ ] Affects user behaviors
- [ ] Contains syntax changes
- [ ] Contains variable changes
- [ ] Contains experimental features
- [ ] Changes MySQL compatibility

### Release note

```release-note
None
```

### CODE REVIEW

File: docs/design/2022-03-10-backup-with-batch-create-table.md [hunk 1]
```diff
@@ -0,0 +1,117 @@
...
+
+## Table of Contents
+
+* [Introduction](#introduction)
+* [Motivation or Background](#motivation-or-background)
+* [Detailed Design](#detailed-design)
+* [Test Design](#test-design)
+ * [Functional Tests](#functional-tests)
+ * [Compatibility Tests](#compatibility-tests)
+ * [Benchmark Tests](#benchmark-tests)
```
Comment (xhebox) [inline docs/design/2022-03-10-backup-with-batch-create-table.md:15]: 1. Some extra spaces here.
2. Lack the section of functional tests. If there is no, remove it and clarify. Or you need to cover it

File: docs/design/2022-03-10-backup-with-batch-create-table.md [hunk 2]
```diff
@@ -0,0 +1,117 @@
+# TiDB Design Documents
+
+- Author(s): [fengou1](http://github.com/your-github-id)
+- Discussion PR: https://github.com/pingcap/tidb/pull/27036
```
Comment (xhebox) [inline docs/design/2022-03-10-backup-with-batch-create-table.md:4]: This is an issue.

File: docs/design/2022-03-10-backup-with-batch-create-table.md [hunk 3]
```diff
@@ -0,0 +1,117 @@
+# TiDB Design Documents
+
+- Author(s): [fengou1](http://github.com/your-github-id)
```
Comment (xhebox) [inline docs/design/2022-03-10-backup-with-batch-create-table.md:3]: Your github id is not modified in the link

File: docs/design/2022-03-10-backup-with-batch-create-table.md [hunk 4]
```diff
@@ -0,0 +1,117 @@
...
+ diff.AffectedOpts = make([]*model.AffectedOption, len(tableInfos))
+ for i := range tableInfos {
+ diff.AffectedOpts[i] = &model.AffectedOption{
+ SchemaID: job.SchemaID,
+ OldSchemaID: job.SchemaID,
+ TableID: tableInfos[i].ID,
+ OldTableID: tableInfos[i].ID,
+ }
+ }
+
```
Comment (xhebox) [inline docs/design/2022-03-10-backup-with-batch-create-table.md:93]: It is too ugly, you need to quote the code with ```

File: docs/design/2022-03-10-backup-with-batch-create-table.md [hunk 5]
```diff
@@ -0,0 +1,117 @@
...
+ RunInTxn(func(txn) {
+ m := meta.New(txn)
+ schemaVesrion := m.CreateTable(t)
+ m.UpdateSchema(schemaVersion)
+ })
+ waitSchemaToSync() // <- This would notify and then 
+ // waiting for all other TiDB nodes are synced with the latest schema version. 
+}
+
+the new design will introduce a new batch create table api
```
Comment (xhebox) [inline docs/design/2022-03-10-backup-with-batch-create-table.md:41]: Introduce more about the logic and the api interface. You don't need to be so detailed in the proposal.

File: docs/design/2022-03-10-backup-with-batch-create-table.md (hunk 1 above)
Comment (fengou1) [reply docs/design/2022-03-10-backup-with-batch-create-table.md]: Fix it.

File: docs/design/2022-03-10-backup-with-batch-create-table.md [hunk 6]
```diff
@@ -0,0 +1,113 @@
...
+## Motivation or Background
+
+The cluster has 6 TiB data, 30k tables, and 11 TiKVs. When I use BR to backup and restore the cluster, The speed is particularly slow. After investigation, BR can only create 2 tables per second, the entire store speed takes nearly 4h, and the execution time spent creating tables is close to 4h. It can be seen that the execution speed of ddl is the bottleneck in this scenario.
+
+## Detailed Design
+
+Backup and Restore for massive tables is extremely slow, the bottleneck of creating the table is to waiting for the schema version change. Each table creates ddl cause a schema version change. 60000 tables have almost 60000 times schema change, it is very cost of restoring massive table
+
+Currently, BR uses an internal interface named CreateTableWithInfo to create table, which creating the table and wait the schema changing one-by-one, omitting the sync of the ddl job between BR and leader, the procedure of creating one table would be like this:
+```
```
Comment (YuJuncen) [inline docs/design/2022-03-10-backup-with-batch-create-table.md:29]: We may fix the indent and set the syntax highlighting.
```suggestion
```go
```

File: docs/design/2022-03-10-backup-with-batch-create-table.md [hunk 7]
```diff
@@ -0,0 +1,113 @@
...
+for _, t := range tables {
+ RunInTxn(func(txn) {
+ m := meta.New(txn)
+ schemaVesrion := m.CreateTable(t)
+ m.UpdateSchema(schemaVersion)
+ })
+```
+
+the new design will introduce a new batch create table api `BatchCreateTableWithInfo`
+```
```
Comment (YuJuncen) [inline docs/design/2022-03-10-backup-with-batch-create-table.md:39]: ditto
```suggestion
```go
```

File: docs/design/2022-03-10-backup-with-batch-create-table.md [hunk 8]
```diff
@@ -0,0 +1,113 @@
...
+
+ for j := range args {
+ if err = d.createTableWithInfoPost(ctx, args[j], jobs.SchemaID); err != nil {
+ return errors.Trace(d.callHookOnChanged(err))
+ }
+ }
+```
+
+For ddl work, introduce a new job type `ActionCreateTables`
+```
```
Comment (YuJuncen) [inline docs/design/2022-03-10-backup-with-batch-create-table.md:74]: ditto
```suggestion
```go
```

**Review by YuJuncen — COMMENTED** [review]
rest lgtm

**Review by xhebox — APPROVED** [review]

**Pull request merged**